Add block hash index which has address states

earlbread opened the pull request:
- Added `IStore.LookupStateReference<T>()` method.
- Added `IStore.StoreStateReference<T>()` method.
- Added `IStore.ForkStateReferences<T>()` method.
- Changed `BlockChain<T>.GetStates()` to use a state reference, which is a `Block<T>.Hash` that has the state of the `Address`.
- Added documentation of `BlockChain<T>.GetStates()`.

dahlia commented on Libplanet/Store/IStore.cs:
```diff
@@ -85,6 +65,59 @@ void PutBlock<T>(Block<T> block, Address addressesMask)
+            where T : IAction, new();
+
+        /// <summary>
+        /// Fork <see cref="Block{T}.Hash"/>es which has the state of the
```
If a sentence in XML comment starts with a verb with a subject omitted it should be third-person singular tense.  See also other existing XML comments.

dahlia commented on Libplanet/Store/IStore.cs:
```diff
@@ -85,6 +65,59 @@ void PutBlock<T>(Block<T> block, Address addressesMask)
+        /// <param name="branchPointIndex">The index of branch point to fork.
+        /// </param>
+        /// <param name="toUpdateAddresses">The set of <see cref="Address"/>es
+        /// to be updated by fork in the <paramref name="targetNamespace"/>.
```
The parameter name and its description seem to neither its intention nor behavior.  It takes a set of addresses that *had been updated* from `branchPointIndex` to the *source*, so its description <q>to be updated by fork in the `targetNamespace`</q> seems wrong and misleading, right?  To me, it looks like addresses to be stripped or invalidated.

earlbread replied:
You're right, I meant <q>addresses to be stripped</q>. I'll fix the parameter name and its description.

dahlia commented on Libplanet/Store/IStore.cs:
```diff
@@ -85,6 +65,59 @@ void PutBlock<T>(Block<T> block, Address addressesMask)
+        /// <param name="address">The <see cref="Address"/> to find the
+        /// <see cref="Block{T}.Hash"/>.</param>
+        /// <param name="offsetIndex">The offset at which to begin looking for
+        /// a block hash.</param>
```
This parameter and `ForkAddressStateBlockHash()` method's `branchPointIndex` takes a block index instead of a block hash, but is there any intention on this?  To me, these parameters can take a block hash and it's less error-prone as block hashes are hard to overlap by mistake (whereas block indexes are easily overlapped and the same index can refer to different blocks depending on branch).

dahlia replied:
Oh, I got it.  These parameters cannot take a block hash because these operations depend on `<=`/`>=` comparisons.

earlbread commented on Libplanet/Store/FileStore.cs:
```diff
@@ -545,6 +532,145 @@ AddressStateMap states
+
+            ImmutableHashSet<Address> updatedAddresses = block.Transactions
+                .SelectMany(tx => tx.UpdatedAddresses)
+                .ToImmutableHashSet();
```
Right, this behavior has to same regardless of implementation. I'll move this to the caller and pass the result as an argument.

earlbread commented on Libplanet/Store/FileStore.cs:
```diff
@@ -545,6 +532,145 @@ AddressStateMap states
+
+            if (!copied)
+            {
+                return;
```
`copied` is `false` when `sourceDirectory` does not exist. I thought if there are no states in the chain that are updated by actions, the `sourceDirectory` may not exist.

dahlia replied:
IMO that case shouldn't be silent no-op but should throw something.

earlbread commented:
I applied review comments and rebased on the master. Also, I renamed methods using a word <q>StateReference</q>. Please take a look. @dahlia @longfin

earlbread commented on Libplanet/Store/FileStore.cs:
```diff
@@ -545,6 +532,141 @@ AddressStateMap states
+
+            if (!copied && addressesToStrip.Any())
+            {
+                throw new DirectoryNotFoundException(
```
Agree, I'll add a new exception for this case.

earlbread replied:
I made `NamespaceNotFoundException`.

dahlia replied:
That's also should be added to the changelog.

dahlia commented on Libplanet/Store/NamespaceNotFoundException.cs:
```diff
@@ -0,0 +1,22 @@
+        /// <see cref="NamespaceNotFoundException"/> class.
+        /// </summary>
+        /// <param name="message">The message that describes the error.</param>
+        public NamespaceNotFoundException(string message)
```
It would be good if this exception has a property to refer to the namespace tried to find.

earlbread commented:
I amended a commit. Please take a look.

dahlia commented on Libplanet/Store/IStore.cs:
```diff
@@ -85,6 +65,80 @@ void PutBlock<T>(Block<T> block, Address addressesMask)
+        /// <param name="address">The <see cref="Address"/> to find lookup.
+        /// </param>
+        /// <param name="lookupFrom">The starting point to lookup a state
+        /// reference. Blocks after <paramref name="lookupFrom"/> are ignored.
```
How about changing the description like: <q>The upper bound (i.e., the latest block) of the search range.  Blocks after `lookupUntil` are ignored.</q>

earlbread replied:
Changed description and parameter name.

dahlia commented on Libplanet/Store/IStore.cs:
```diff
@@ -85,6 +65,80 @@ void PutBlock<T>(Block<T> block, Address addressesMask)
+        /// </param>
+        /// <param name="address">The <see cref="Address"/> to find lookup.
+        /// </param>
+        /// <param name="lookupFrom">The starting point to lookup a state
```
I suggest another parameter name: `lookupUntil`.

dahlia commented on Libplanet/Store/IStore.cs:
```diff
@@ -85,6 +65,80 @@ void PutBlock<T>(Block<T> block, Address addressesMask)
+        /// Forks state references, which are <see cref="Block{T}.Hash"/>es that
+        /// have the state of the <see cref="Address"/>es, from
+        /// <paramref name="sourceNamespace"/> to
+        /// <paramref name="destNamespace"/>.
```
To me `destinationNamespace` looks better.

earlbread replied:
Changed.

longfin commented on Libplanet/Store/BaseStore.cs:
```diff
@@ -52,11 +53,7 @@ ISet<TxId> txids
-
-        /// <inheritdoc />
-        public abstract void PutBlock<T>(Block<T> block, Address addressesMask)
+        public abstract void PutBlock<T>(Block<T> block)
```
```suggestion
        /// <inheritdoc />
        public abstract void PutBlock<T>(Block<T> block)
```

longfin commented on Libplanet/Blockchain/BlockChain.cs:
```diff
@@ -472,6 +466,25 @@ internal BlockChain<T> Fork(HashDigest<SHA256> point)
+
+                var toUpdateAddresses = new HashSet<Address>();
+
+                for (var i = pointBlock.Index + 1; i <= Tip.Index; i++)
```
(minor) We can omit block index lookup by looping reverse order using `Tip.PreviousHash` and `Blocks[Hash]`.

earlbread commented:
I rebased on the master and applied review comments. Please take a look.

dahlia commented on Libplanet/Blockchain/BlockChain.cs:
```diff
@@ -472,6 +466,28 @@ internal BlockChain<T> Fork(HashDigest<SHA256> point)
+
+                for (
+                    Block<T> block = Tip;
+                    block.PreviousHash is HashDigest<SHA256> hash
```
This condition won't work if `block` refers to a block right after genesis, which means addresses updated by the second block (where `blockIndex == 1`) are always excluded from the index.

earlbread replied:
I think this condition is break when the block refers to the genesis block, not the second block. But I thought it is ok because every forking must have the genesis block.
I added tests for this case and found another bug and fixed it.

earlbread commented:
I have fixed the part that uses `Transaction.UpdatedAddress`. Because it is for validation, it may be different from the result of `BlockChain<T>.SetStates`.

longfin commented on Libplanet/Blockchain/BlockChain.cs:
```diff
@@ -359,14 +345,19 @@ bool render
 
                     Store.UnstageTransactionIds(txIds);
+
+                    ImmutableHashSet<Address> updatedAddresses =
```
Is this variable needed yet?

earlbread replied:
No, it doesn't. I removed it.

longfin commented:
> I have fixed the part that uses `Transaction.UpdatedAddress`. Because it is for validation, it may be different from the result of `BlockChain<T>.SetStates`.

It would be also better if there is a regression test about it.

longfin commented:
> > I have fixed the part that uses `Transaction.UpdatedAddress`. Because it is for validation, it may be different from the result of `BlockChain<T>.SetStates`.
> 
> It would be also better if there is a regression test about it.

I've missed. it's added 👍
